API.md

```diff
@@ -1,6 +1,96 @@
-# Authmosphere API documentation
+# Authmosphere API Documentation
 
-## Mock tooling
+## Express Tooling
+
+Authmosphere provides two middleware factories to secure Express based http services.
+
+### authenticationMiddleware
+
+Middleware that handles OAuth authentication for API endpoints. It extracts and validates the `access token` from the request.
+
+If configured as a global middleware(see usage section), all request need to provide a valid token to access the endpoint.
+<br>
+If some endpoints should be excluded from this restriction, they need to be added to the `options.publicEndpoints` array to be whitelisted.
+
+If validation of the provided token fails the middleware rejects the request with status _401 UNAUTHORIZED_. <br>
+To overwrite this behavior a custom handler can be specified by passing in `options.onNotAuthenticatedHandler` (see [`onNotAuthenticatedHandler`](./src/types/AuthenticationMiddlewareOptions.ts)).
+
+* ⚠️&nbsp;&nbsp;While this middleware could also be configured per endpoint (i.e. `app.get(authenticationMiddleware(...), endpoint)` it is not recommended as using it as global middleware will force you into a whitelist setup.
+  * Make sure `authenticationMiddleware` is at the top of the registered request handlers. This is essential to gurantee the enforceability of the whitelist strategy.
+* ⚠️&nbsp;&nbsp;The middleware attaches metadata (scopes of the token) to the express request object. The `requireScopesMiddleware` relies on this information.
+
+
+#### Usage
+
+```typescript
+import {
+  authenticationMiddleware
+} from 'authmosphere';
+
+app.use(authenticationMiddleware({
+  publicEndpoints: ['/heartbeat', '/status'],
+  tokenInfoEndpoint: 'auth.example.com/tokeninfo'
+});
+```
+
+#### Signature
+
+`authenticationMiddleware(options) => express.RequestHandler`
+
+#### Arguments
+
+* [`options`](./src/types/AuthenticationMiddlewareOptions.ts):
+  * `tokenInfoEndpoint: string` - url of the Token validation endpoint
+  * `publicEndpoints?: string[] - list of whitelisted API paths`
+  * `logger?: Logger` - [logger](./src/types/Logger.ts)
+  * [`onNotAuthenticatedHandler?: onNotAuthenticatedHandler`](./src/types/AuthenticationMiddlewareOptions.ts) - custom response handler
+
+### requireScopesMiddleware
+
```

Comment:
Suggestion:

> A factory that returns a middleware that compares scopes attached to `express.Request` object with a given list (`scopes` parameter). If all required scopes are matched, the middleware calls `next`. Otherwise, it rejects the request with _403 FORBIDDEN_.

Reply:
✅